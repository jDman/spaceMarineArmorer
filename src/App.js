import React, { Fragment, useCallback, useEffect, useState } from 'react';
import { Switch, Redirect, Route } from 'react-router-dom';
import axios from 'axios';
import { withRouter, useHistory } from 'react-router-dom';

import './App.css';
import Header from './components/global/Header';
import Nav from './components/global/Nav';
import ArmorAdmin from './pages/admin/ArmorAdmin';
import ArmorShop from './pages/shop/ArmorShop';
import Cart from './pages/cart/Cart';
import Signup from './pages/auth/Signup';
import Login from './pages/auth/Login';

function App() {
  let history = useHistory();

  const [isAuth, setAuth] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [generatedToken, setGeneratedToken] = useState();

  const addToCartHandler = (event, armor) => {
    event.preventDefault();

    const armorId = armor.id;
    const quantity = armor.config.value;

    return axios
      .put('http://localhost:4000/shop/armor/cart', {
        armorId,
        quantity
      })
      .then(result => {
        console.log(result);
        return result;
      })
      .catch(err => {
        console.log(err);
      });
  };

  const deleteCartItemHandler = itemId => {
    return axios
      .delete(`http://localhost:4000/shop/armor/cart/item`, {
        params: { itemId }
      })
      .then(result => result.message)
      .catch(err => console.log(err));
  };

  const loadingHandler = loadingValue => {
    setIsLoading(loadingValue);
  };

  const loginHandler = (event, { loading, loginFormIsValid, loginForm }) => {
    event.preventDefault();

    loadingHandler(true);

    axios
      .post('http://localhost:4000/auth/login', {
        email: loginForm.email.value,
        password: loginForm.password.value
      })
      .then(result => {
        const remainingMilliseconds = 60 * 60 * 1000;
        const token = result.data.token;

        const expiryDate = new Date(
          new Date().getTime() + remainingMilliseconds
        );

        sessionStorage.setItem('expiryDate', expiryDate.toISOString());
        sessionStorage.setItem('token', token);
        sessionStorage.setItem('userId', result.data.userId);

        setAuth(true);
        setGeneratedToken(token);

        return;
      })
      .then(() => {
        history.replace('/shop');
      })
      .catch(err => {
        console.log(err);
      })
      .finally(() => {
        loadingHandler(false);
      });
  };

  const logoutHandler = useCallback(() => {
    setAuth(false);

    sessionStorage.removeItem('expiryDate');
    sessionStorage.removeItem('token');
    sessionStorage.removeItem('userId');

    history.replace('/login');
  }, [history]);

  const signupHandler = (event, { loading, signupFormIsValid, signupForm }) => {
    event.preventDefault();

    loadingHandler(true);

    axios
      .post('http://localhost:4000/auth/signup', {
        userName: signupForm.userName.value,
        email: signupForm.email.value,
        password: signupForm.password.value
      })
      .then(result => {
        history.replace('/login');
      })
      .catch(err => {
        console.log(err);
      })
      .finally(() => {
        loadingHandler(false);
      });
  };

  const orderItemsHandler = items => {
    return axios
      .post(`http://localhost:4000/shop/armor/order`, {
        items
      })
      .then(result => result.message)
      .catch(err => console.log(err));
  };

  useEffect(() => {
    const token = sessionStorage.getItem('token');
    const expiryDate = sessionStorage.getItem('expiryDate');
    const expired = new Date(expiryDate) <= new Date();

    if (expired) {
      logoutHandler();
      return;
    }

    setAuth(true);
    setGeneratedToken(token);
  }, [logoutHandler]);

  return (
    <Fragment>
      <div className="App">
        <Header />
        <Nav isAuth={isAuth} logout={logoutHandler} />
      </div>

      {isLoading ? <h1>Loading ...</h1> : null}

      <Switch>
        <Route path="/admin" render={props => <ArmorAdmin />}></Route>

        <Route
          path="/cart"
          render={props => {
            if (generatedToken) {
              return (
                <Cart
                  token={generatedToken}
                  deleteCartItem={deleteCartItemHandler}
                  orderItems={orderItemsHandler}
                />
              );
            }
          }}
        ></Route>

        <Route
          path="/shop"
          render={() => {
            if (generatedToken) {
              return (
                <ArmorShop
                  addToCart={addToCartHandler}
                  token={generatedToken}
                  loading={loadingHandler}
                  isLoading={isLoading}
                />
              );
            }
          }}
        ></Route>

        <Route
          path="/signup"
          render={props => <Signup signup={signupHandler} />}
        ></Route>

        <Route
          path="/login"
          render={props => <Login login={loginHandler} />}
        ></Route>

        <Route path="/">
          <p>Home</p>
        </Route>

        <Redirect to="/" />
      </Switch>
    </Fragment>
  );
}

export default withRouter(App);
