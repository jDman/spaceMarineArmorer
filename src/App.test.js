import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter as Router } from 'react-router-dom';

import App from './App';

describe('<App />', () => {
  it('renders App component', () => {
    const component = renderer.create(
      <Router>
        <App />
      </Router>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
