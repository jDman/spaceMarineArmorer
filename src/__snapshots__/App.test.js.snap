// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders App component 1`] = `
Array [
  <div
    className="App"
  >
    <header>
      <h1>
        Space Soldier Armory
      </h1>
    </header>
    <nav>
      <ul
        className="Navigation"
      >
        <li
          className="NavigationListItem"
        >
          <a
            aria-current={null}
            className="NavigationLink"
            href="/signup"
            onClick={[Function]}
          >
            Signup
          </a>
        </li>
        <li
          className="NavigationListItem"
        >
          <a
            aria-current="page"
            className="NavigationLink active"
            href="/login"
            onClick={[Function]}
            style={
              Object {
                "color": "#3300ff",
              }
            }
          >
            Login
          </a>
        </li>
      </ul>
    </nav>
  </div>,
  <section
    className="auth-form"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="Input"
      >
        <label
          className="Label"
        >
          email
        </label>
        <input
          className="InputElement"
          name="email"
          onChange={[Function]}
          placeholder="you@email.com"
          required={true}
          type="email"
          value=""
        />
      </div>
      <div
        className="Input"
      >
        <label
          className="Label"
        >
          password
        </label>
        <input
          className="InputElement"
          name="password"
          onChange={[Function]}
          placeholder="Enter you password..."
          required={true}
          type="password"
          value=""
        />
      </div>
      <button
        type="submit"
      >
        Login
      </button>
    </form>
  </section>,
]
`;
